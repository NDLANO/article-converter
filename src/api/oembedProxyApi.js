/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import fetch from 'isomorphic-fetch';
import { apiResourceUrl, resolveJsonOrRejectWithError, headerWithAccessToken } from '../utils/apiHelpers';

export const fetchOembed = (embed, accessToken) =>
  fetch(apiResourceUrl(`/oembed-proxy/v1/oembed?url=${embed.url}`), { headers: headerWithAccessToken(accessToken) })
    .then(resolveJsonOrRejectWithError)
    .then(oembed => ({ ...embed, oembed }))
    .catch((e) => {
      if (e.status === 501 && e.json.code === 'PROVIDER_NOT_SUPPORTED') {
        return { ...embed, message: `Uhåndtert embed med følgende url: ${embed.url}`, resource: 'error' };
      }
      return { ...embed, message: `En uventet feil oppsto med følgende embed url: ${embed.url}`, resource: 'unhandled-oembed-error' };
    });
