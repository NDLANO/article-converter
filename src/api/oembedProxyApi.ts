/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import fetch from 'isomorphic-fetch';
import queryString from 'query-string';
import {
  apiResourceUrl,
  resolveJsonOrRejectWithError,
  headerWithAccessToken,
} from '../utils/apiHelpers';
import { EmbedType } from '../interfaces';
import { OembedEmbedType } from '../plugins/externalPlugin';

export interface OembedProxyResponse {
  type: string;
  version: string;
  title?: string;
  description?: string;
  authorName?: string;
  authorUrl?: string;
  providerName?: string;
  providerUrl?: string;
  cacheAge?: number;
  thumbnailUrl?: string;
  thumbnailWidth?: number;
  thumbnailHeight?: number;
  width?: number;
  height?: number;
  html?: string;
}

export const fetchOembed = async (
  embed: EmbedType,
  accessToken: string,
): Promise<OembedEmbedType> => {
  const url = new URL(typeof embed.data.url === 'string' ? embed.data.url : '');
  if (url.hostname.includes('youtu') && url.protocol === 'http:') {
    url.protocol = 'https:';
    embed.data.url = url.href;
  }
  const response = await fetch(
    apiResourceUrl(
      `/oembed-proxy/v1/oembed?${queryString.stringify({
        url: embed.data.url,
      })}`,
    ),
    {
      headers: headerWithAccessToken(accessToken),
    },
  );
  const oembed = await resolveJsonOrRejectWithError<OembedProxyResponse>(response);
  return {
    ...embed,
    oembed,
  };
};
