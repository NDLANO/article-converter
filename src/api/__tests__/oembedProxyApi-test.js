/**
 * Copyright (c) 2021-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import nock from 'nock';
import { fetchOembed } from '../oembedProxyApi';

test('Converts youtube http-url to https', async () => {
  nock('http://ndla-api')
    .get('/oembed-proxy/v1/oembed?url=https://youtube.com/videoid')
    .reply(200, {
      html: '<iframe src="https://youtube.com/videoid"></iframe>',
    });

  const oembed = {
    data: {
      url: 'http://youtube.com/videoid',
    },
  };
  const result = await fetchOembed(oembed);
  expect(result).toMatchSnapshot();
});

test('Leaves youtube https-url be', async () => {
  nock('http://ndla-api')
    .get('/oembed-proxy/v1/oembed?url=https://youtube.com/anotherid')
    .reply(200, {
      html: '<iframe src="https://youtube.com/anotherid"></iframe>',
    });

  const oembed = {
    data: {
      url: 'https://youtube.com/anotherid',
    },
  };
  const result = await fetchOembed(oembed);
  expect(result).toMatchSnapshot();
});
