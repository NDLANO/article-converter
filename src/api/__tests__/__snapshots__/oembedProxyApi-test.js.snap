// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Converts youtube http-url to https 1`] = `
Object {
  "data": Object {
    "url": "https://youtube.com/videoid",
  },
  "oembed": Object {
    "html": "<iframe src=\\"https://youtube.com/videoid\\"></iframe>",
  },
}
`;

exports[`Leaves youtube https-url be 1`] = `
Object {
  "data": Object {
    "url": "https://youtube.com/anotherid",
  },
  "oembed": Object {
    "html": "<iframe src=\\"https://youtube.com/anotherid\\"></iframe>",
  },
}
`;
